lib/bn/bn128_arithmetic.ex

```diff
@@ -0,0 +1,153 @@
+    cond do
+      !on_curve?(point1, b) ->
+        {:error, "point1 is not on the curve"}
+
+      !on_curve?(point2, b) ->
+        {:error, "point2 is not on the curve"}
+
+      true ->
+        {:ok, add_points(point1, point2)}
+    end
+  end
+
+  @spec mult(Point.t(), integer(), integer()) :: {:ok, Point.t()} | {:error, String.t()}
+  def mult(point, scalar, b \\ @default_b) do
+    if !on_curve?(point, b) do
+      {:error, "point1 is not on the curve"}
+    else
+      {:ok, mult_point(point, scalar)}
+    end
+  end
+
+  @spec mult_point(Point.t(), integer()) :: Point.t()
+  defp mult_point(point, scalar) do
+    cond do
+      scalar == 0 ->
+        {:ok, result} = Point.new(0, 0, modulus: point.modulus)
+
+        result
+
+      scalar == 1 ->
+        point
+
+      rem(scalar, 2) == 0 ->
+        point
+        |> mult_point(div(scalar, 2))
+        |> double()
+
+      true ->
+        point
+        |> mult_point(div(scalar, 2))
+        |> double()
+        |> calculate_points_addition(point)
+    end
+  end
+
+  @spec add_points(Point.t(), Point.t()) :: Point.t()
+  defp add_points(point1, point2) do
+    cond do
+      point1 == point2 ->
+        double(point1)
+
+      infinity?(point1) ->
+        point2
+
+      infinity?(point2) ->
+        point1
+
+      true ->
+        calculate_points_addition(point1, point2)
+    end
+  end
+
+  @spec double(Point.t()) :: Point.t()
+  defp double(point) do
+    if point.y.value == 0 do
+      {:ok, result} = Point.new(0, 0, modulus: point.modulus)
+
+      result
+    else
+      double_y = IntegerModP.mult(point.y, 2)
+
+      lambda =
+        point.x
+        |> IntegerModP.pow(2)
+        |> IntegerModP.mult(3)
+        |> IntegerModP.div(double_y)
+
+      double_x = IntegerModP.mult(point.x, 2)
+
+      x = lambda |> IntegerModP.pow(2) |> IntegerModP.sub(double_x)
+
+      y =
+        point.x
+        |> IntegerModP.sub(x)
+        |> IntegerModP.mult(lambda)
+        |> IntegerModP.sub(point.y)
+
+      %Point{
+        x: x,
+        y: y,
+        modulus: point.modulus
+      }
+    end
+  end
+
+  @spec calculate_points_addition(Point.t(), Point.t()) :: Point.t()
+  defp calculate_points_addition(point1, point2) do
+    if point1.x == point2.x do
+      {:ok, result} = Point.new(0, 0, modulus: point1.modulus)
+
+      result
+    else
+      y_remainder = IntegerModP.sub(point2.y, point1.y)
+      x_remainder = IntegerModP.sub(point2.x, point1.x)
+      lambda = IntegerModP.div(y_remainder, x_remainder)
+
+      x =
+        lambda
+        |> IntegerModP.pow(2)
+        |> IntegerModP.sub(point1.x)
+        |> IntegerModP.sub(point2.x)
+
+      y =
+        point1.x
+        |> IntegerModP.sub(x)
+        |> IntegerModP.mult(lambda)
+        |> IntegerModP.sub(point1.y)
+
+      %Point{
+        x: x,
+        y: y,
+        modulus: point1.modulus
+      }
+    end
+  end
+
+  @spec infinity?(Point.t()) :: boolean()
+  def infinity?(point) do
```

Comment:
curious: why is `(0,0)` called `infinity` not `zero`?

Reply:
@masonforest  it's described in eip196
`Curve points are encoded as two field elements (x, y), where the point at infinity is encoded as (0, 0).`